### ISSUE TITLE
Low level SFX API proposal

### ISSUE THREAD

**Issue by dxxb** [moduscreate/ATMlib#5]
Features:
* Global state (in SRAM) is 6 bytes
* Provides to _grab_ multiple channels for SFX rendering in a single call
* Simple and very flexible so application specific utility code or API sugar can be built on top of it

Limitations:
* Only one sound effect can be rendered at a time
* SFX duration limited to maximum of 0xFFFF ticks (it should affect anything)

``` C
/* Only one sound effect at a time. SFX duration limited to maximum of 0xFFFF ticks */

/* SFX rendering callback

 Once the channel(s) have been _grabbed_ this function can alter
 fields in the struct osc associated with the grabbed channel(s).

Rules:
 Called synchronously on SFX start with ticks_counter = 0.
 It must grab channels with asfx_grab_channels() before using them.
 Called synchronously on SFX stop with ticks_counter = 0xFFFF.
 Must call asfx_reset_fx() when done or when called with ticks_counter = 0xFFFF.
*/
typedef void(asfx_fx_callback*)(void *priv, uint16_t ticks_counter);

struct asfx_state {
	uint16_t asfx_ticks_counter;
	asfx_fx_callback asfx_current_callback;
	void *priv;
};

extern struct asfx_state;

#define ASFX_CH0_MASK (1)
#define ASFX_CH1_MASK (1<<1)
#define ASFX_CH2_MASK (1<<2)
#define ASFX_CH3_MASK (1<<3)

/* Grab channels from ATMlib for SFX rendering

Music playback is muted on those channels and the effect rendering callback
has full control of the corresponding OSC.
*/
void asfx_grab_channels(uint8_t channels_mask);

/* Set the call-back invocation rate in Hz

This can be called at any time in the application or in the effect rendering callback.
*/
void asfx_set_callback_rate(uint16_t cb_rate_hz);

/* Reset FX callback info when finished rendering an effect

This function must be called from the callback when it is done rendering the effect
or when the callback is called with a tick_counter value of 0xFFFF.
*/
void asfx_reset_fx(void);

/* Start rending an effect using the specified callback and private param pointer

This call is the main entry point for SFX rendering. An application calls this function
providing a callback and an optional private data pointer which will be passed
to the callback.

The callback is called immediately before this function returns with ticks_counter
set to zero to give the callback a chance to setup the channels and tick rate as required.
After that the callback will be invoked at the rate set by asfx_set_callback_rate() with
an incrementing value of ticks_counter until the callback signals it is done by calling
asfx_reset_fx() or the application calls asfx_stop_fx().

If this function is called while an effect is being rendered the current effect callback is
called with a ticks_counter value of 0xFFFF before the new effect callback is setup.
*/
void asfx_start_fx(asfx_fx_callback cb, void *priv);

/* Signal the sound effect callback to stop

If any sound effect is being rendered the relevant callback is called one last time
with a ticks_counter value of 0xFFFF. The callback is expected to run asfx_stop_fx()
and return.
*/
void asfx_stop_fx(void);
```

## Example usage: low level API

``` C
void laser_sfx_callback(void *priv, uint16_t ticks_counter)
{
    /* setup and teardown for similar SFX can be factored out to reduce code size */
    if (!ticks_counter) {
        asfx_grab_channels(ASFX_CH1_MASK);
        asfx_set_callback_rate(35);
        osc[1].vol = 63;
        return;
    } else if (ticks_counter == 0xFFFF) {
        asfx_reset_fx();
        return;
    }

    /* render the effect */
    osc[1].phase_increment = 300+ticks_counter;
}

void somefunc(void)
{
    ....
    asfx_start_fx(laser_sfx_callback, NULL);
    ....
    /* Optional SFX stop. The FX callback should stop after it is done */
    asfx_stop_fx();
}

```

## Example usage: start SFX by index

``` C
/* level 1 sugar: SFX index */

enum sfx_e {
	SFX_BOOM = 0,
	SFX_LASERFIRE,
};

struct sfx_def {
	asfx_fx_callback sfx_callback;
	void *priv;
}  sfx_array[] {
	{boom_cb, NULL},
	{laserfire_cb, NULL},
};

#define play_sfx_by_index(idx) asfx_start_fx(sfx_array[idx].sfx_callback, sfx_array[idx].priv)

void somefunc(void)
{
    ....
    play_sfx_by_index(SFX_LASERFIRE);
    ....
}
```

## Example usage: build a tracker like SFX renderer

``` C
/* sugar level: OVER 9000 */

struct sfx_tracker_state {
    uint8_t *pos;
    ...
} sfx_tracker;

void sfx_tracker_callback(void *priv, uint16_t ticks_counter)
{
    struct sfx_tracker_state *s = priv;
    ...
    /* do the tracker thing here */
}

void play_sfx_track(uint8_t *sfx_track)
{
    sfx_tracker.pos = sfx_track;
    ...
    asfx_start_fx(sfx_tracker_callback, &sfx_tracker);
}

PROGMEM uint8_t laserfire[] = {0x02, ...};

void somefunc(void)
{
    ....
    play_sfx_track(laserfire);
    ....
}
```

**Comment by jaygarcia** [comment]
This looks really good.

Questions:

- I like the idea of the callback pattern, but feel that it might 1) be too much overhead (this is an assumption) and 2) too complex for folks who are not experts in C (or understand what a callback is).
- I like the idea of using a data structure to define SFX sequences.  What do you think about this rather simple spec?

```
const unsigned char PROGMEM  pewPew[] = {
      0x01, // Channel
      0x02, // Num Sequences
      0xF9, 0x10, // Frequency, Milliseconds
      0x13F, 0x1F0 // Frequency, Milliseconds
};

playSfx(pewPew);
```

@dxxb =)

**Comment by dxxb** [comment]
> I like the idea of the callback pattern, but feel that it might 1) be too much overhead

Callbacks, threads or [coroutines](https://github.com/renaudbedard/littlebits-arduino/tree/master/Libraries/Coroutines) are the options for making execution asynchronous. Callbacks is the solution with less CPU and memory overhead of the three. Threads are out of the question and coroutines are awesome but I would not push them because:
1. it takes up more program space than the callback solution
2. I haven't tested coroutines on the atmega and given we don't have much time I am wary of adding more untested dependencies.

> 2) too complex for folks who are not experts in C (or understand what a callback is)

Callbacks are annoying but are widely used in GUI toolkits and until recently very common in JavaScript too. But I agree with you because ...

> I like the idea of using a data structure to define SFX sequences.

... I like the idea of using a data structure too. Your example is in the same vain as my 3rd example above. Basically this proposal is for a barebone API which provides the most flexible mechanism while imposing minimum policy on its users (this is an ideal goal for any library). The tracker-like SFX play routine can be built on top on the API and if the tracker-like API covers all needs the user need not know about the low level one. But having the low level one enables you to achieve SFX which expressed in the simple tracker format of your example would take up more program space.

To recap: this is a low level API proposal, users don't need to be using it directly. Yes, we should look at some tracker-like byte stream format for generating SFX and build it on top of the low level API.

**Comment by jaygarcia** [comment]
Loving how this is turning out.

**Comment by dxxb** [comment]
Glad you like it! So armed with the above basic API idea I looked into how to make work with the existing interrupt handler and tracker playroutine. My conclusion was that by having a _per channel_ tick rate counter and a _per channel_ tick handler we can achieve two things:
1. different tempo per channel (the tempo command becomes per channel instead of global) and,
2. SFX can take advantage of the existing callback mechanism but with their own tick rate.

This also paves the way for the next step: using the same track format for music and SFX (less code to write and more free program memory, YAY!).

**Comment by dxxb** [comment]
Per-channel tempo is IN 51b6c3b6e55a72fdbe0cce5ad22b201f058270d8 (and a bunch of other changes) 😉

**Comment by jaygarcia** [comment]
@dxxb Having different tempos per channel can produce a problem for music.  For the composition side, we only have to describe tempo once, versus four times. That's an additional 4byte overhead for songs that only need tempo set globally. Trackers all work on a global tempo.

**Comment by dxxb** [comment]
@jaygarcia, thanks for the comment and I agree. I reached the same conclusion since pushing that commit and now went back to have a global song tick rate plus an independent SFX tick rate.

**Comment by dxxb** [comment]
I have something working. I went straight to the pattern SFX solution because I figured that's what you prefer and hope to use for the game. There is time to add the simple API above later.

I have the following pattern:
``` C
struct mod_sfx_state sfx_state;

const PROGMEM struct sfx1_data {
  uint8_t num_tracks; /* this can go away */
  uint16_t tracks_offset[2]; /* we can make these offsets bytes for SFX patterns */
  uint8_t pattern1[8];
  uint8_t pattern2[9];
} sfx1 = {
  .num_tracks = 0x02,
  .tracks_offset = {
      offsetof(struct sfx1_data, pattern1),
      offsetof(struct sfx1_data, pattern2),
  },
  .pattern1 = {
    0x9D, 10,  
    0x40, 10,
    0xFD, 10, 1,
    0x9F,       
  },
  .pattern2 = {
    0x00 + 42,    // NOTE ON: note = 42
    0x9F + 20,   // DELAY: ticks = 20
    0x00 + 43,    // NOTE ON: note = 43
    0x9F + 20,   // DELAY: ticks = 20
    0x00 + 44,    // NOTE ON: note = 44
    0x9F + 20,   // DELAY: ticks = 20
    0x00 + 45,    // NOTE ON: note = 45
    0x9F + 20,   // DELAY: ticks = 20
    0xFE,         // RETURN
  }
};
```

Triggered (and re-triggered) or stopped by:
``` C
 if (arduboy.justPressed(B_BUTTON)) {
    atm_synth_play_sfx_track(CH_ONE, (struct mod_sfx*)&sfx1, &sfx_state);
  }

  if (arduboy.justPressed(A_BUTTON)) {
    atm_synth_stop_sfx_track(&sfx_state);
  }
```

**Comment by jaygarcia** [comment]
This looks *REALLY* good. 

Now that each track has it's own tempo, we can use this pattern to do cool shit like use FX (glissando, portamento (slides), etc.).
